opera/rules_test.go

```diff
@@ -198,3 +202,40 @@ func TestRules_MinBaseFee_NoCopy_PreAllegro(t *testing.T) {
 		t.Errorf("original and copied rules must be the same - shallow copy for preAllegro: got %d, want %d", got, want)
 	}
 }
+
+func TestCreateTransientEvmChainConfig(t *testing.T) {
```

Comment:
There should be at least one test dealing with more than one entry in the `[]UpgradeHeight` slice. Also, one with an empty input list should be added to make clear what the expectation in such a case is.

Since this function essentially becomes the canonical interpretation of the `[]UpgradeHeight`, it should also be tested how the `heights` in the array are interpreted. If an upgrade is listed as being send in for block X, is it valid already for block X or for block X+1?